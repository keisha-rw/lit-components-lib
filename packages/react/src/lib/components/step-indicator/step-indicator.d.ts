import * as React from 'react';
export declare const PdsStepIndicator: React.FunctionComponent<Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
