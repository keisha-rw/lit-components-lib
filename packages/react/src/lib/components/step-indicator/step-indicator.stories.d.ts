import * as React from 'react';
declare const _default: {
    title: string;
    component: React.FunctionComponent<Partial<{
        variant: "horizontal" | "vertical";
        inverted: boolean;
        interactive: boolean;
    }> & {
        ref?: React.Ref<any> | undefined;
    } & {
        style?: React.CSSProperties | undefined;
    } & Partial<{}> & {
        id?: string | undefined;
        slot?: string | undefined;
        className?: string | undefined;
        children?: React.ReactNode;
    }>;
    tags: string[];
};
export default _default;
export declare const Horizontal: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const HorizontalInteractive: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const Vertical: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const VerticalInteractive: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const HorizontalInverted: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const HorizontalInteractiveInverted: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "horizontal" | "vertical";
    inverted: boolean;
    interactive: boolean;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const PreviousStepActive: {
    (): JSX.Element;
    storyName: string;
};
export declare const PreviousStepActiveInverted: {
    (): JSX.Element;
    storyName: string;
    parameters: {
        backgrounds: {
            default: string;
        };
    };
};
