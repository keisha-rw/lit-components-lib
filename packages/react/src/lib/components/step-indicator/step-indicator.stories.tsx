import * as React from 'react';
import { StoryFn } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { PdsStepIndicatorItem } from '../step-indicator-item/step-indicator-item';
import { PdsStepIndicator } from './step-indicator';

export default {
  title: 'Components/Step indicator',
  component: PdsStepIndicator,
  tags: ['autodocs'],
};

const Template: StoryFn<typeof PdsStepIndicator> = (args) => {
  if (args.interactive) {
    return (
      <PdsStepIndicator {...args}>
        <PdsStepIndicatorItem
          onClick={(e: any) =>
            action('pds-step-indicator-item-click')(e.detail)
          }
          status="completed"
          href="#"
        >
          Step 1
        </PdsStepIndicatorItem>
        <PdsStepIndicatorItem
          status="current"
          active
          href="#"
          onClick={(e: any) => {
            e.preventDefault();
            // eslint-disable-next-line no-console
            console.log(e);
          }}
        >
          Step two is longer and wraps to the second line
        </PdsStepIndicatorItem>
        <PdsStepIndicatorItem
          onClick={(e: any) =>
            action('pds-step-indicator-item-click')(e.detail)
          }
          status="incomplete"
        >
          Step 3
        </PdsStepIndicatorItem>
        <PdsStepIndicatorItem
          onClick={(e: any) =>
            action('pds-step-indicator-item-click')(e.detail)
          }
          status="incomplete"
        >
          Step 4
        </PdsStepIndicatorItem>
      </PdsStepIndicator>
    );
  }

  return (
    <PdsStepIndicator {...args}>
      <PdsStepIndicatorItem
        onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
        status="completed"
      >
        Step 1
      </PdsStepIndicatorItem>
      <PdsStepIndicatorItem
        onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
        status="current"
        active
      >
        Step two is longer and wraps to the second line
      </PdsStepIndicatorItem>
      <PdsStepIndicatorItem
        onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
        status="incomplete"
      >
        Step 3
      </PdsStepIndicatorItem>
      <PdsStepIndicatorItem
        onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
        status="incomplete"
      >
        Step 4
      </PdsStepIndicatorItem>
    </PdsStepIndicator>
  );
};

export const Horizontal = Template.bind({});
Horizontal.args = {};

export const HorizontalInteractive = Template.bind({});
HorizontalInteractive.storyName = 'Horizontal interactive';
HorizontalInteractive.args = {
  interactive: true,
};

export const Vertical = Template.bind({});
Vertical.args = {
  variant: 'vertical',
};

export const VerticalInteractive = Template.bind({});
VerticalInteractive.storyName = 'Vertical interactive';
VerticalInteractive.args = {
  interactive: true,
  variant: 'vertical',
};

export const HorizontalInverted = Template.bind({});
HorizontalInverted.storyName = 'Horizontal inverted';
HorizontalInverted.args = {
  inverted: true,
};
HorizontalInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

export const HorizontalInteractiveInverted = Template.bind({});
HorizontalInteractiveInverted.storyName = 'Horizontal interactive inverted';
HorizontalInteractiveInverted.args = {
  interactive: true,
  inverted: true,
};
HorizontalInteractiveInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};

export const PreviousStepActive = () => (
  <PdsStepIndicator>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="completed"
      href="#"
      active
    >
      Step 1
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="current"
      href="#"
    >
      Step two is longer and wraps to the second line
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="incomplete"
    >
      Step 3
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="incomplete"
    >
      Step 4
    </PdsStepIndicatorItem>
  </PdsStepIndicator>
);
PreviousStepActive.storyName = 'Previous step active';

export const PreviousStepActiveInverted = () => (
  <PdsStepIndicator inverted>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="completed"
      href="#"
      active
    >
      Step 1
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="current"
      href="#"
    >
      Step two is longer and wraps to the second line
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="incomplete"
    >
      Step 3
    </PdsStepIndicatorItem>
    <PdsStepIndicatorItem
      onClick={(e: any) => action('pds-step-indicator-item-click')(e.detail)}
      status="incomplete"
    >
      Step 4
    </PdsStepIndicatorItem>
  </PdsStepIndicator>
);
PreviousStepActiveInverted.storyName = 'Previous step active inverted';
PreviousStepActiveInverted.parameters = {
  backgrounds: { default: 'BrandXStrong' },
};
