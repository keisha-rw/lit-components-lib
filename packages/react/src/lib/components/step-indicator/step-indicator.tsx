import * as React from 'react';
import { PdsStepIndicator as PdsStepIndicatorElement } from '@keisha/design-system';
import { PdsStepIndicatorProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsStepIndicator =
  createPdsReactComponent<PdsStepIndicatorProps>()({
    tagName: 'pds-step-indicator',
    elementClass: PdsStepIndicatorElement,
    react: React,
  });
