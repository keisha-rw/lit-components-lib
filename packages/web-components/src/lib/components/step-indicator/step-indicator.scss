/* stylelint-disable string-no-newline */
@import '../../../style-utils/svg';

/*------------------------------------*\
 #STEP_INDICATOR
\*------------------------------------*/

:host {
  display: block;
}

.pds-c-step-indicator {
  // setting local css variables for this component so we can control
  // the step indicator items based on props on this parent component

  /** Text colors */
  --pds-step-indicator-item-text-completed: var(--pds-semantic-text-subtle);
  --pds-step-indicator-item-text-current: var(--pds-semantic-text-default);
  --pds-step-indicator-item-text-incomplete: var(--pds-semantic-text-subtle);
  --pds-step-indicator-item-text-interactive: var(
    --pds-semantic-text-interactive-default
  );
  --pds-step-indicator-item-text-interactive-hover: var(
    --pds-semantic-text-interactive-strong
  );
  --pds-step-indicator-item-text-interactive-active: var(
    --pds-semantic-text-interactive-xstrong
  );

  /** Line colors */
  --pds-step-indicator-item-after-completed-border: var(
    --pds-semantic-border-interactive-default
  );
  --pds-step-indicator-item-after-incomplete-border: var(
    --pds-semantic-border-default
  );

  /** Circle icons */

  // completed
  --pds-step-indicator-item-circle-completed-bg: var(
    --pds-semantic-background-interactive-default
  );
  --pds-step-indicator-item-circle-completed-bg-hover: var(
    --pds-semantic-background-interactive-strong
  );
  --pds-step-indicator-item-circle-completed-bg-active: var(
    --pds-semantic-background-interactive-xstrong
  );
  --pds-step-indicator-item-circle-completed-check: var(
    --pds-semantic-border-inverted-default
  );

  // current
  --pds-step-indicator-item-circle-current-border: var(
    --pds-semantic-border-interactive-default
  );
  --pds-step-indicator-item-circle-current-bg: var(
    --pds-semantic-background-interactive-default
  );
  --pds-step-indicator-item-circle-current-border-hover: var(
    --pds-semantic-border-interactive-strong
  );
  --pds-step-indicator-item-circle-current-bg-hover: var(
    --pds-semantic-background-interactive-strong
  );
  --pds-step-indicator-item-circle-current-border-active: var(
    --pds-semantic-border-interactive-xstrong
  );
  --pds-step-indicator-item-circle-current-bg-active: var(
    --pds-semantic-background-interactive-xstrong
  );

  // incomplete
  --pds-step-indicator-item-circle-incomplete-border: var(
    --pds-semantic-border-default
  );

  /** Layout */
  --pds-step-indicator-item-direction: row;
  --pds-step-indicator-item-align: top;
  --pds-step-indicator-item-step-margin: 0;
  --pds-step-indicator-item-circle-top: 2px; // magic number to align circle with text
}

.pds-c-step-indicator--inverted {
  /** Text colors */
  --pds-step-indicator-item-text-completed: var(
    --pds-semantic-text-inverted-default
  );
  --pds-step-indicator-item-text-current: var(
    --pds-semantic-text-inverted-default
  );
  --pds-step-indicator-item-text-incomplete: var(
    --pds-semantic-text-inverted-default
  );
  --pds-step-indicator-item-text-interactive: var(
    --pds-semantic-text-inverted-default
  );
  --pds-step-indicator-item-text-interactive-hover: var(
    --pds-semantic-text-interactive-inverted-strong
  );
  --pds-step-indicator-item-text-interactive-active: var(
    --pds-semantic-text-interactive-inverted-xstrong
  );

  /** Line colors */
  --pds-step-indicator-item-after-completed-border: var(
    --pds-semantic-border-inverted-default
  );
  --pds-step-indicator-item-after-incomplete-border: var(
    --pds-semantic-border-inverted-default
  );

  /** Circle icons */

  // completed
  --pds-step-indicator-item-circle-completed-bg: var(
    --pds-semantic-background-default
  );
  --pds-step-indicator-item-circle-completed-bg-hover: var(
    --pds-semantic-background-interactive-inverted-strong
  );
  --pds-step-indicator-item-circle-completed-bg-active: var(
    --pds-semantic-background-interactive-inverted-xstrong
  );
  --pds-step-indicator-item-circle-completed-check: var(
    --pds-semantic-border-interactive-default
  );

  // current
  --pds-step-indicator-item-circle-current-border: var(
    --pds-semantic-border-inverted-default
  );
  --pds-step-indicator-item-circle-current-bg: var(
    --pds-semantic-background-default
  );
  --pds-step-indicator-item-circle-current-border-hover: var(
    --pds-semantic-border-interactive-inverted-strong
  );
  --pds-step-indicator-item-circle-current-bg-hover: var(
    --pds-semantic-background-interactive-inverted-strong
  );
  --pds-step-indicator-item-circle-current-border-active: var(
    --pds-semantic-border-interactive-inverted-xstrong
  );
  --pds-step-indicator-item-circle-current-bg-active: var(
    --pds-semantic-background-interactive-inverted-xstrong
  );

  // incomplete
  --pds-step-indicator-item-circle-incomplete-border: var(
    --pds-semantic-border-inverted-default
  );
}

.pds-c-step-indicator__list {
  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 0;
  list-style: none;

  .pds-c-step-indicator--horizontal & {
    @media all and (min-width: $pds-bp-lg) {
      flex-direction: row;
      justify-content: space-between;
    }
  }
}

// apply the connector line to all the items, expect the last
::slotted(pds-step-indicator-item:not(:last-child)) {
  position: relative;

  &::after {
    content: '';
    position: absolute;
    // magic numbers to dotted align the line correctly
    top: 27px;
    left: 15.5px;
    // 100% minus the size of the circle with 1px overlap
    height: calc(100% - 16px);
    width: 3px;
    background-position: 50% 50%;
    background-repeat: repeat;
    background-image: todatauri(
      '<svg viewBox="0 0 20 20" height="8" width="8"> \
        <circle r="3" cx="10" cy="10" fill="#c5c6c5" /> \
      </svg>'
    );

    .pds-c-step-indicator--inverted & {
      background-image: todatauri(
        '<svg viewBox="0 0 20 20" height="8" width="8"> \
          <circle r="3" cx="10" cy="10" fill="#ffffff" /> \
        </svg>'
      );
    }
  }
}

::slotted(pds-step-indicator-item[status='completed']) {
  &::after {
    // magic number to align the line correctly
    left: 16px;
    width: 0;
    border-color: var(--pds-step-indicator-item-after-completed-border);
    border-style: solid;
    border-width: 0 0 0 2px;
  }
}

.pds-c-step-indicator--horizontal {
  @media all and (min-width: $pds-bp-lg) {
    /** Layout */
    --pds-step-indicator-item-direction: column;
    --pds-step-indicator-item-align: center;
    --pds-step-indicator-item-step-margin: auto;
    --pds-step-indicator-item-circle-top: 0;

    ::slotted(pds-step-indicator-item) {
      // horizontal line
      &::after {
        // magic number to align the dotted line correctly
        top: 16px;
        // half way plus a litle more for the circle
        left: calc(50% + 4px);
        height: 3px;
        // 100% minus the size of the circle with 1px overlap
        width: calc(100% - 16px);
      }
    }

    ::slotted(pds-step-indicator-item[status='completed']) {
      &::after {
        // magic numbers to align the solid line correctly
        top: 16.5px;
        height: 0;
        border-width: 2px 0 0;
      }
    }
  }
}
