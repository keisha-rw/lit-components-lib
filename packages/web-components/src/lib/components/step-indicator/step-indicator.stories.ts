import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './step-indicator';
import '../step-indicator-item/step-indicator-item';

export default {
  title: 'Components/Step indicator',
  component: 'pds-step-indicator',
  tags: ['autodocs'],
  parameters: {
    a11y: {
      config: {
        rules: [
          {
            // Color contrast is sufficient for the incomplete checks, but axe doesn't like them.
            //
            // @link https://dequeuniversity.com/rules/axe/4.6/color-contrast
            id: 'color-contrast',
            enabled: false,
          },
        ],
      },
    },
  },
  argTypes: {
    variant: {
      control: 'select',
      options: ['horizontal', 'vertical'],
    },
  },
  render: (args) => {
    if (args.interactive) {
      return html`<pds-step-indicator
        variant="${args.variant || nothing}"
        ?interactive="${args.interactive}"
        ?inverted="${args.inverted}"
      >
        <pds-step-indicator-item status="completed" href="#">
          Step one
        </pds-step-indicator-item>
        <pds-step-indicator-item status="current" href="#" active>
          Step two is longer and wraps to the second line
        </pds-step-indicator-item>
        <pds-step-indicator-item status="incomplete">
          <span>Step three</span>
        </pds-step-indicator-item>
        <pds-step-indicator-item status="incomplete">
          <span>Step four</span>
        </pds-step-indicator-item>
      </pds-step-indicator>`;
    }

    return html`<pds-step-indicator
      variant="${args.variant || nothing}"
      ?interactive="${args.interactive}"
      ?inverted="${args.inverted}"
    >
      <pds-step-indicator-item status="completed">
        <span>Step one</span>
      </pds-step-indicator-item>
      <pds-step-indicator-item status="current" active>
        <span>Step two is longer and wraps to the second line</span>
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step three</span>
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step four</span>
      </pds-step-indicator-item>
    </pds-step-indicator>`;
  },
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Horizontal: StoryObj = {
  name: 'Horizontal',
  args: {},
};

export const HorizontalInteractive: StoryObj = {
  name: 'Horizontal interactive',
  args: {
    interactive: true,
  },
  parameters: {
    actions: {
      handles: ['pds-step-indicator-item-click'],
    },
  },
};

export const Vertical: StoryObj = {
  name: 'Vertical',
  args: {
    variant: 'vertical',
  },
};

export const VerticalInteractive: StoryObj = {
  name: 'Vertical interactive',
  args: {
    variant: 'vertical',
    interactive: true,
  },
  parameters: {
    actions: {
      handles: ['pds-step-indicator-item-click'],
    },
  },
};

export const HorizontalInverted: StoryObj = {
  name: 'Horizontal inverted',
  args: {
    inverted: true,
  },
  parameters: {
    backgrounds: { default: 'BrandXStrong' },
  },
};

export const HorizontalInteractiveInverted: StoryObj = {
  name: 'Horizontal interactive inverted',
  args: {
    interactive: true,
    inverted: true,
  },
  parameters: {
    backgrounds: { default: 'BrandXStrong' },
    actions: {
      handles: ['pds-step-indicator-item-click'],
    },
  },
};

export const PreviousStepActive: StoryObj = {
  name: 'Previous step active',
  parameters: {
    actions: {
      handles: ['pds-step-indicator-item-click'],
    },
  },
  render() {
    return html`<pds-step-indicator variant="horizontal" interactive>
      <pds-step-indicator-item status="completed" href="#" active>
        Step one
      </pds-step-indicator-item>
      <pds-step-indicator-item status="current" href="#">
        Step two is longer and wraps to the second line
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step three</span>
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step four</span>
      </pds-step-indicator-item>
    </pds-step-indicator>`;
  },
};

export const PreviousStepActiveInverted: StoryObj = {
  name: 'Previous step active inverted',
  parameters: {
    backgrounds: { default: 'BrandXStrong' },
    actions: {
      handles: ['pds-step-indicator-item-click'],
    },
  },
  render() {
    return html`<pds-step-indicator variant="horizontal" interactive inverted>
      <pds-step-indicator-item status="completed" href="#" active>
        Step one
      </pds-step-indicator-item>
      <pds-step-indicator-item status="current" href="#">
        Step two is longer and wraps to the second line
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step three</span>
      </pds-step-indicator-item>
      <pds-step-indicator-item status="incomplete">
        <span>Step four</span>
      </pds-step-indicator-item>
    </pds-step-indicator>`;
  },
};
