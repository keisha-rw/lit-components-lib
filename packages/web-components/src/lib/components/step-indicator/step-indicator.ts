import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { msg, localized } from '@lit/localize';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import '../step-indicator-item/step-indicator-item';
import styles from './step-indicator.scss?inline';

/**
 * @summary A step indicator
 *
 * @slot @slot default Contains the step indicator items, restricted to pds-step-indicator-item element
 */
@customElement('pds-step-indicator', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
@localized()
export class PdsStepIndicator extends PdsElement {
  connectedCallback() {
    super.connectedCallback();
    this.setLocale();
  }

  /**
   * variant
   * - **horizontal** renders the step-indicator horizontal on large screens and up
   * - **vertical** renders the step-indicator always vertical
   */
  @property()
  variant: 'horizontal' | 'vertical' = 'horizontal';

  /**
   * interactive
   */
  @property({ type: Boolean })
  interactive: boolean = false;

  /**
   * inverted
   */
  @property({ type: Boolean })
  inverted: boolean = false;

  /**
   * @internal
   */
  @queryAssignedElements({
    slot: undefined,
    selector: 'pds-step-indicator-item',
  })
  stepIndicatorItems: HTMLElement[];

  /**
   * @internal
   * pushes properties down to subcomponent step indicator items
   */
  addPropertiesToStepIndicatorItems() {
    this.stepIndicatorItems.forEach((stepIndicatorItem) => {
      stepIndicatorItem.setAttribute('inverted', '');
    });
  }

  handleSlotChange(e: Event) {
    this.handleSlotValidation(e);
    this.addPropertiesToStepIndicatorItems();
  }

  /**
   * @internal
   */
  get classNames() {
    return {
      [this.variant]: !!this.variant,
      inverted: !!this.inverted,
    };
  }

  listTemplate() {
    return html`<ol role="list" class="${this.classEl('list')}">
      <slot
        allowed-elements="pds-step-indicator-item"
        @slotchange=${this.handleSlotChange}
        >Step indicator items</slot
      >
    </ol>`;
  }

  render() {
    if (this.interactive) {
      return html`<nav
        class="${this.getClass()}"
        aria-label="${msg('progress')}"
      >
        ${this.listTemplate()}
      </nav>`;
    }

    return html`<div
      class="${this.getClass()}"
      role="group"
      aria-label="${msg('progress')}"
    >
      ${this.listTemplate()}
    </div>`;
  }
}
