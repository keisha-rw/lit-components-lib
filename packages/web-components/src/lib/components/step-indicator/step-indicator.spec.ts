import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsStepIndicator } from './step-indicator';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Step indicator/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('StepIndicator unit tests', () => {
  let element: Element;

  beforeEach(async () => {
    element = await fixture(
      '<pds-step-indicator> <pds-step-indicator-item> </pds-step-indicator-item> </pds-step-indicator>',
    );
  });

  it('is an instance of PdsStepIndicator', () => {
    expect(element).toBeInstanceOf(PdsStepIndicator);
  });

  it('renders a <nav> when interactive', async () => {
    element = await fixture('<pds-step-indicator interactive />');
    expect(element.shadowRoot?.querySelector('nav')).toBeTruthy();
  });

  it('renders Spanish aria label when interactive', async () => {
    element = await fixture('<pds-step-indicator interactive lang="es"/>');
    expect(
      element.shadowRoot?.querySelector('nav')?.getAttribute('aria-label'),
    ).toBe('progreso');
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('populates the step indicator item slot when it is type of pds-step-indicator-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture(
      `<pds-step-indicator variant="horizontal" interactive>
        <pds-step-indicator-item status="completed" href="#">
          Step one
        </pds-step-indicator-item>
      </pds-step-indicator>`,
    );
    expect(consoleError).not.toHaveBeenCalled();
    consoleError.mockRestore();
  });

  it('does not populate the step indicator item slot when it is not type of pds-step-indicator-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture(
      `<pds-step-indicator variant="horizontal" interactive>
      <div status="completed" href="#">
        Step one
      </div>
    </pds-step-indicator>`,
    );
    expect(element).toMatchSnapshot();
    expect(consoleError).toBeCalledTimes(1);
    consoleError.mockRestore();
  });
});
