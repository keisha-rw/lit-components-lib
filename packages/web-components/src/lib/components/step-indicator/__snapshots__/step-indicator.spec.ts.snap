// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepIndicator unit tests does not populate the step indicator item slot when it is not type of pds-step-indicator-item 1`] = `
<pds-step-indicator
  interactive=""
  variant="horizontal"
>
  
      
  
    
</pds-step-indicator>
`;
